### ISSUE TITLE
iOS13: Xamarin.Forms SearchBar BackgroundColor not behaving properly

### ISSUE THREAD

**Issue by baileysnead** [xamarin/Xamarin.Forms#8041]
### Description
Since iOS13 update, setting the BackgroundColor property of the Xamarin.Forms SearchBar sets both the search bar border and the text-field fill to the same color. I want these two aspects to be different colors for contrast reasons.

### Steps to Reproduce

`<SearchBar BackgroundColor="Gray"></SearchBar>`

### Expected Behavior
Outer box to search bar is gray, inner text field is white
![Screen Shot 2019-10-16 at 9 38 46 AM](https://user-images.githubusercontent.com/56640895/66924773-39f89b80-eff9-11e9-845e-fe0d009b7dc9.png)

### Actual Behavior
Both outer box to search bar and text field are both gray
![Screen Shot 2019-10-16 at 9 44 11 AM](https://user-images.githubusercontent.com/56640895/66924931-88a63580-eff9-11e9-97e1-65f37d4a554e.png)

### Basic Information

- Version with issue: iOS13, Visual Studio 8.3.4, XCode 11
- Last known good version: Before iOS13
- IDE: Visual Studio

**Comment by adrianknight89** [comment]
I've noticed this (or something similar) recently when testing the gallery app on iOS13.

**Comment by renzska** [comment]
I am experiencing this same issue. I used an Effect to modify the SearchBar to have a white background in the meantime (stuck it in a border effect that I had created temporarily).

UISearchBar.SearchTextField.BackgroundColor = UIColor.White;

**Comment by MathewLC** [comment]
I can confirm that when I put BackgroundColor=White or BackgroundColor=#FFFFFF, the expected behavior as describe by @baileysnead occur. Perphaps, SearchTextField.BackgroundColor should be exposed as a iOS Platform-specific property. I would like to give a try on doing that if that change be accepted.

**Comment by justinhorner** [comment]
Sorry for the wall of text.

Also experencing the same issue described by @baileysnead.

Also tried a custom renderer and having no luck there either.

```
using System.ComponentModel;
using Foundation;
using App.iOS.Renderers;
using UIKit;
using Xamarin.Forms;
using Xamarin.Forms.Platform.iOS;

[assembly: ExportRenderer(typeof(SearchBar), typeof(CustomSearchBarRenderer))]

namespace App.iOS.Renderers
{
    public class CustomSearchBarRenderer : SearchBarRenderer
    {
        protected override void OnElementChanged(ElementChangedEventArgs<SearchBar> e)
        {
            base.OnElementChanged(e);

            //!!! Works, but only for the first search bar (we have 3) **
            UITextField.AppearanceWhenContainedIn(typeof(UISearchBar)).BackgroundColor =
            UIColor.FromRGB(255, 0, 0);

            //Match text field within SearchBar to its background color
            using (var searchKey = new NSString("_searchField"))
            {
                if (e.NewElement == null) return;

                //!!! Throws an iOS error on iOS 13 ***
                var textField = (UITextField)Control.ValueForKey(searchKey);
                textField.BackgroundColor = e.NewElement.BackgroundColor.ToUIColor();
                textField.TintColor = UIColor.White;
            }
        }

        protected override void OnElementPropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            base.OnElementPropertyChanged(sender, e);

            // Hide Cancel Button
            if (e.PropertyName == "Text") Control.ShowsCancelButton = false;
        }
    }
}
```

** This line works but only for the first searchbar. We have 3 search bars (on different tabs...not sure if that is a cause) and only the first search bar utilizes this code
*** This throws the below error when debugging on iOS 13+

```
Foundation.MonoTouchException: 
  at at (wrapper managed-to-native) ObjCRuntime.Messaging.xamarin_IntPtr_objc_msgSend_IntPtr(intptr,intptr,intptr)
  at Foundation.NSObject.ValueForKey (Foundation.NSString key) [0x0001c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/13.2.0.47/src/Xamarin.iOS/NSObject.g.cs:602
  at App.iOS.Renderers.CustomSearchBarRenderer.OnElementChanged (Xamarin.Forms.Platform.iOS.ElementChangedEventArgs`1[TElement] e) [0x00024] in /Users/user-name/Documents/App/iOS/Renderers/CustomSearchBarRenderer.cs:22
  at Xamarin.Forms.Platform.iOS.VisualElementRenderer`1[TElement].SetElement (TElement element) [0x0012a] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementRenderer`1[TElement].Xamarin.Forms.Platform.iOS.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00000] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.Platform.CreateRenderer (Xamarin.Forms.VisualElement element) [0x00014] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementPackager.OnChildAdded (Xamarin.Forms.VisualElement view) [0x0003d] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementPackager.Load () [0x0001e] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementRenderer`1[TElement].SetElement (TElement element) [0x000de] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementRenderer`1[TElement].Xamarin.Forms.Platform.iOS.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00000] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.Platform.CreateRenderer (Xamarin.Forms.VisualElement element) [0x00014] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementPackager.OnChildAdded (Xamarin.Forms.VisualElement view) [0x0003d] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementPackager.Load () [0x0001e] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementRenderer`1[TElement].SetElement (TElement element) [0x000de] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementRenderer`1[TElement].Xamarin.Forms.Platform.iOS.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00000] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.Platform.CreateRenderer (Xamarin.Forms.VisualElement element) [0x00014] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementPackager.OnChildAdded (Xamarin.Forms.VisualElement view) [0x0003d] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementPackager.Load () [0x0001e] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementRenderer`1[TElement].SetElement (TElement element) [0x000de] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementRenderer`1[TElement].Xamarin.Forms.Platform.iOS.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00000] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.Platform.CreateRenderer (Xamarin.Forms.VisualElement element) [0x00014] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementPackager.OnChildAdded (Xamarin.Forms.VisualElement view) [0x0003d] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementPackager.Load () [0x0001e] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementRenderer`1[TElement].SetElement (TElement element) [0x000de] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementRenderer`1[TElement].Xamarin.Forms.Platform.iOS.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00000] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.Platform.CreateRenderer (Xamarin.Forms.VisualElement element) [0x00014] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementPackager.OnChildAdded (Xamarin.Forms.VisualElement view) [0x0003d] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementPackager.Load () [0x0001e] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementRenderer`1[TElement].SetElement (TElement element) [0x000de] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementRenderer`1[TElement].Xamarin.Forms.Platform.iOS.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00000] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.Platform.CreateRenderer (Xamarin.Forms.VisualElement element) [0x00014] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementPackager.OnChildAdded (Xamarin.Forms.VisualElement view) [0x0003d] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Platform.iOS.VisualElementPackager.OnChildAdded (System.Object sender, Xamarin.Forms.ElementEventArgs e) [0x0000f] in <fc33cd67c9774c61ba6dca5e7bdd7329>:0
  at Xamarin.Forms.Element.OnChildAdded (Xamarin.Forms.Element child) [0x0000f] in D:\a\1\s\Xamarin.Forms.Core\Element.cs:312
  at Xamarin.Forms.VisualElement.OnChildAdded (Xamarin.Forms.Element child) [0x00000] in D:\a\1\s\Xamarin.Forms.Core\VisualElement.cs:755
  at Xamarin.Forms.Layout.OnInternalAdded (Xamarin.Forms.View view) [0x0001d] in D:\a\1\s\Xamarin.Forms.Core\Layout.cs:453
  at Xamarin.Forms.Layout.InternalChildrenOnCollectionChanged (System.Object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x00078] in D:\a\1\s\Xamarin.Forms.Core\Layout.cs:443
  at System.Collections.ObjectModel.ObservableCollection`1[T].OnCollectionChanged (System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x00018] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/corefx/src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs:263
  at System.Collections.ObjectModel.ObservableCollection`1[T].OnCollectionChanged (System.Collections.Specialized.NotifyCollectionChangedAction action, System.Object item, System.Int32 index) [0x00000] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/corefx/src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs:338
  at System.Collections.ObjectModel.ObservableCollection`1[T].InsertItem (System.Int32 index, T item) [0x0001a] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/corefx/src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs:196
  at System.Collections.ObjectModel.Collection`1[T].Add (T item) [0x00020] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/external/corefx/src/Common/src/CoreLib/System/Collections/ObjectModel/Collection.cs:71
  at Xamarin.Forms.ContentPresenter.OnContentChanged (Xamarin.Forms.BindableObject bindable, System.Object oldValue, System.Object newValue) [0x0004e] in D:\a\1\s\Xamarin.Forms.Core\ContentPresenter.cs:88
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) [0x00000] in /Users/builder/jenkins/workspace/xamarin-macios/xamarin-macios/external/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1021
  at Foundation.NSAsyncSynchronizationContextDispatcher.Apply () [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/13.2.0.47/src/Xamarin.iOS/Foundation/NSAction.cs:178
  ```

  The solution we found, for now, is to only run the code textField.BackgroundColor code (***) when the iOS version is less than 13.  This ensures the search bar looks okay for iOS 13+ and anything lower.  Not ideal but it works

**Comment by MathewLC** [comment]
Taking the @renzska response and on this [stackoverflow response](https://stackoverflow.com/a/58324928/11305148) I was able to use the effect and works fine. The following code is from the link above:
```
public class SearchBarBackgroundEffect : PlatformEffect
{
        public SearchBarBackgroundEffect()
        {
        }

        private UIKit.UISearchBar NativeSearchBar => (UIKit.UISearchBar)Control;
        private SearchBar XamarinSearchBar => (SearchBar)Element;

        protected override void OnAttached()
        {
            if (UIKit.UIDevice.CurrentDevice.CheckSystemVersion(13, 0))
                NativeSearchBar.SearchTextField.BackgroundColor = XamarinSearchBar.BackgroundColor.ToUIColor();
        }
        protected override void OnDetached()
        {

        }
    }

**Comment by attrib75** [comment]
@MathewLC interesting, the stack overflow solution you linked does not work for me.  I tried it.  The background color is still white

**Comment by attrib75** [comment]
HI @MathewLC, thanks for the help.  I'll try to throw something togethr

**Comment by attrib75** [comment]
@justinhorner After seeing your post I thought I might give it a try:
 UITextField.AppearanceWhenContainedIn(typeof(UISearchBar)).BackgroundColor =
            UIColor.FromRGB(255, 0, 0);

It worked for me as well, but it only seems to change the first instance of a search bar it encounters in the app.  Very strange behavior.

**Comment by attrib75** [comment]
@MathewLC I created a repo of the problem here
[https://github.com/attrib75/XamIoSSearchBarColorBug](https://github.com/attrib75/XamIoSSearchBarColorBug)
thanks!

**Comment by MathewLC** [comment]
@justinhorner I downloaded your repo and made the test placing the searchBar on all pages. It worked for me. The only thing that is happen is that for some reason I'm changing the xaml and is not refleteced on iOS build and to get done I had to clean, close visual studio and open again. Also I think I found a bug or I doing something wrong but removing the "x:Name" property of the search Bar in first page, cause app to crash.
Ignoring all that, using just the Background Color and and removing the name(x:Name property) of the search Bar on another pages, works fine. If you still having trouble on that, you can update your repo with the version that only gets done in the first instance of searchBar. For now, maybe later today I'll make my repo and share with you.
> @MathewLC I created a repo of the problem here
> https://github.com/attrib75/XamIoSSearchBarColorBug
> thanks!

**Comment by attrib75** [comment]
Hello, the sample works for you as is?  I don't understand.  The searchbar background is white on iOS 13 with the sample I uploaded.

**Comment by MathewLC** [comment]
@attrib75  Yeah, your sample worked for me, the only thing I did was remove placeholder color as the background color is applied to both background and placeholder background.
> @MathewLC
> Hello, the sample works for you as is? I don't understand. The searchbar background is white on iOS 13 with the sample I uploaded.

This is what it seems like:
![image](https://user-images.githubusercontent.com/37350787/67860387-77325280-fafc-11e9-85d9-a392864b8126.png)
![image](https://user-images.githubusercontent.com/37350787/67860409-84e7d800-fafc-11e9-96a2-7f40a1da73a6.png)
![image](https://user-images.githubusercontent.com/37350787/67860422-8e714000-fafc-11e9-9306-697615ca2ecf.png)

This is in simulator anyway, I can't tell if something change in device or how much difference is your environment from mine.

**Comment by attrib75** [comment]
@MathewLC I have the iPhone 11 Pro Max 13.0 and it is white.

**Comment by MathewLC** [comment]
I can't try in a device right now but as soon as I can and  the bug appears I'll try a solution, if you find first thought, let me now, please.
> @MathewLC I have the iPhone 11 Pro Max 13.0 and it is white.

**Comment by attrib75** [comment]
Sorry I meant earlier I have the iPhone 11 Pro Max 13.0 _simulator_ whereas yours is 13.1

**Comment by attrib75** [comment]
attempting upgrade of xcode

**Comment by attrib75** [comment]
Great news!  It works after updating XCode, Xamarin and VS.  For others with this issue in this thread, try it

**Comment by MathewLC** [comment]
@attrib75 after so much job you finally found out. Congratulations, I'm sure many others will enjoy find this issue. For now, I'm just waiting xamarin team approval to try minimize the work with some platform specific property.

**Comment by attrib75** [comment]
I spoke too soon. It works in the sample repo but my app still shows white :(.  Not sure what the difference is.

**Comment by attrib75** [comment]
Ahh I see the problem, it depends on the simulator used

iPhone 11 - 13.1:  Doesn't work
iPhone 11 Pro Max - 13.1 Works

**Comment by MathewLC** [comment]
I'm not sure that is it.
![image](https://user-images.githubusercontent.com/37350787/67892274-2427c200-fb33-11e9-96f7-39260d88ee51.png)
But my version of xamarin,forms is not the last one, when you updated, do you update to the last one? Maybe that's it?
> Ahh I see the problem, it depends on the simulator used
> 
> iPhone 11 - 13.1: Doesn't work
> iPhone 11 Pro Max - 13.1 Works

**Comment by attrib75** [comment]
I solved it now (see edited comment above) thanks!

**Comment by klogeaage** [comment]
I think the SearchBar needs some attention to make it work properly again on iOS 13.x. If you look carefully at the original image of this issue, you will notice that it is not only the color that is wrong, it also misses the Cancel button that used to be there in previous iOS versions.

**Comment by MathewLC** [comment]
@klogeaage I think that if the right properties can be exposed to we can use them as we need to, and adjust the appearance according the the application needs.

**Comment by klogeaage** [comment]
Agreed, but it would also be nice if the out-of-the-box behavior is sensible.

**Comment by pierre01** [comment]
I have the same issue and nothing has been done to fix it... it makes my product look unprofessional. if a property is exposed why isn't it applied on the first time?
In these examples the correct properties are applied after navigating around the app...
![SearchBox01](https://user-images.githubusercontent.com/2603287/76140042-52e70500-600b-11ea-89e2-683e0df566f3.jpg)
![SearchBox02](https://user-images.githubusercontent.com/2603287/76140048-5da19a00-600b-11ea-93ed-fd6bdb802b36.jpg)

**Comment by LeoJHarris** [comment]
Still a problem..

**Comment by jfversluis** [comment]
I think we fixed this in #13804! Let me know if that's not the case :)

**Comment by stephanpalmer** [comment]
Hi @jfversluis, correct me if I'm wrong, but I think the original bug report is not addressed. The original bug report is related to the non-shell SearchBar view. Unless I'm mistaken, fix #13804 is Shell specific and does not apply to the non-shell SearchBar view (the change class `SearchHandlerAppearanceTracker` is not used in `SearchBarRenderer`).

To my understanding, the original bug report is related to the fact that the default background color of the `UISearchTextField` contained in `UISearchBar` used to be white up to an including iOS 12. Starting with iOS 13, the background color is a semi-transparent grey. This causes the SearchBar's background color to shine through the search text entry field. This is also the reason why the fixes proposed in the following comments above work: [Comment 1](https://github.com/xamarin/Xamarin.Forms/issues/8041#issuecomment-544658047), [Comment 2](https://github.com/xamarin/Xamarin.Forms/issues/8041#issuecomment-546022694). So this issue is really related to Apple's changes to the UISearchBar appearance.

When comparing the most recent Xamarin.Forms version 5.0.0.2291 (SR8) with 5.0.0.2083 (SR4), the behavior seems unchanged. With both version, the SearchBar looks identical. Here are some example screenshots from different iOS versions:

### iOS 11.4

![ios11 4](https://user-images.githubusercontent.com/4060315/148531207-25a6dbea-593b-4f34-916e-40353ef5a7cc.png)

### iOS 13.7 

![ios13 7](https://user-images.githubusercontent.com/4060315/148531268-827750f3-b06f-4ea8-9ea7-66a39558dee2.png)

### iOS 15.0

![ios15 0](https://user-images.githubusercontent.com/4060315/148531283-efe1fc47-7c63-47d4-abd7-42f5636fc011.png)

**Comment by jfversluis** [comment]
@stephanpalmer interesting. I think you're totally right :)

What would be the right course of action here according to you?

It seems that iOS had an update and our controls followed along accordingly. From what I understand you would want to background to remain white, right? Or rather; the right color according to light/dark theme.

However others might see this as a UI change that comes with a new version of iOS and welcome it. 

Simply resetting the background color to white (taking in regard dark mode as well) doesn't seem like an option then. That would only retain Forms behavior but doesn't then reflect iOS platform behavior anymore.

**Comment by stephanpalmer** [comment]
@jfversluis Thank you for the update. The expectation stated in the original bug report is that the background color of the search text field remains white on iOS 13+. That was also my assumption when writing my previous comment.

However, I agree with your assessment that there probably is no default behaviour which is correct for all Xamarin.Forms users. Some will want the control to have the "iOS default look", others will want to customise the background colors of both the search bar and the search text field. In addition, as you have mentioned, there is light and dark mode to consider.

So I think what we need is the possibility to set the background color for the search bar and the search text field independently. If this is easily possible with iOS specific features like the appearance API, we probably do not need an Xamarin.Forms API for this.

I've tested the fix you proposed using the appearance API. It works great on iOS 14 and 15, but does not work on iOS 13 (tested with 13.6 and 13.7 specifically) for me. I so far did not find out why iOS 13 does not work.

Setting the search text field background color in a custom renderer or effect works on iOS 13, 14 and 15.

Regarding the formal resolution of this issue, I would not consider it fixed in #13804, since this code change had no effect on this issue. But based on your assessment, the issue is probably not a fixable regression. Instead, a good resolution would be to determine a stable and preferred way to control the iOS search bar and search text field colors independently on iOS 13+.
